Update relationships.rst (#93)
autofac/Documentation: docs/resolve/relationships.rst

If you would rather control ``B`` disposal yourself all the time, :doc:`register B as ExternallyOwned() <../lifetime/disposal>`.

> Dynamic Instantiation (`Func<B>`)

Reviewer: RST uses double backticks for code, but they don't work well in headers.
```suggestion
Dynamic Instantiation (Func<B>)
```

-------------------------------

> Using an *auto-generated factory* can let you effectively call ``Resolve<B>()`` without tying your component to Autofac. Use this relationship type if you need to create more than one instance of a given service, or if you're not sure if you're going to need a service and want to make the decision at runtime. This relationship is also useful in cases like :doc:`WCF integration <../integration/wcf>` where you need to create a new service proxy after faulting the channel.

Author: Suggested rephrasing of 'preceding sentence'. See conversation above.

```suggestion
Using an *auto-generated factory* can let you resolve a new `B` programmatically within the control flow of your program, without requiring a direct dependency on the Autofac Library. Use this relationship type if:
 * you need to create more than one instance of a given service
 * you want to specifically control when the setup of the service occurs
 * you're not sure if you're going to need a service and want to make the decision at runtime.
This relationship is also useful in cases like :doc:`WCF integration <../integration/wcf>` where you need to create a new service proxy after faulting the channel.
```

Reviewer: This looks good. My slightly-nitpicky technical writing classes from college are kicking in, though, so some minor changes to make the bullets consistent and not use "Autofac Library" as a proper noun.

```suggestion
Using an *auto-generated factory* can let you resolve a new `B` programmatically within the control flow of your program, without requiring a direct dependency on the Autofac library. Use this relationship type if:

* You need to create more than one instance of a given service.
* You want to specifically control when the setup of the service occurs.
* You're not sure if you're going to need a service and want to make the decision at runtime.

This relationship is also useful in cases like :doc:`WCF integration <../integration/wcf>` where you need to create a new service proxy after faulting the channel.
```

> Note that whilst this could be used to simply invoke a parameterless constructor of the target type ``B``, that isn't all it's limited to. As stated above, invoking the ``Func<B>`` is more akin to calling ``Resolve<B>``, and hence Autofac can wire up other types passed into the constructor, as long as it knows how to provide those types. Hence this calling format can be used for any type where all the ctor dependencies can be resolved by the container.

Author: Splitting existing suggestion across multiple lines

```suggestion
``Func<B>`` behaves just like calling ``Resolve<B>()``. That means it's not limited to acting on parameterless constructors on objects - it will wire up constructor parameters, do property injection, and follow the whole lifecycle that ``Resolve<B>()`` does.
```

Reviewer: Looks good. [I'm not sure if I commit the suggestion or you do; there's a button allowing me to do that but I'm a bit scared of where that's going to go. I'll let you do it. ;) ]

> Another corollary of this behaving like calling ``Resolve<B>()`` is that **Lifetime scopes are respected** using this relationship type. If you register an object as ``InstancePerDependency()`` and call the ``Func<B>`` multiple times, you'll get a new instance each time. However, if you register an object as ``SingleInstance()`` and call the ``Func<B>`` to resolve the object more than once, you will get *the same object instance every time*.

Reviewer: I see the clarification here is that `Func<B>` is not just for parameterless constructors - it's identical in almost every respect to calling `Resolve<B>`. I'm curious if we can be just slightly more succinct in explaining that.

[It doesn't seem suggestions work on multiple lines, at least in the preview. This is a suggestion for the 149-151 lines.]

Maybe something like...

```suggestion
``Func<B>`` behaves just like calling ``Resolve<B>()``. That means it's not limited to acting on parameterless constructors on objects - it will wire up constructor parameters, do property injection, and follow the whole lifecycle that ``Resolve<B>()`` does.

Further, lifetime scopes are respected. If you register an object as ``InstancePerDependency()`` and call the ``Func<B>`` multiple times, you'll get a new instance each time; if you register an object as ``SingleInstance()`` and call the ``Func<B>`` to resolve the object more than once, you will get *the same object instance every time*.
```

Author: I like that, but the first sentence of the suggestion is now repeating the start of the preceding sentence. I've suggested a change to that preceding sentence to make it more about intent that implementation, to remove that repetition.

Author: Issue with Multi-Line suggestions is irritating, isn't it. :(
I've created single-line suggestions applying your changes, so that we can stick within this review workflow. If you're happy with the adjustments to the preceding sentence, then I'll commit all the changes and Resolve this Conversation.

Reviewer: Yeah, that's cool.

I filed a complaint about the multiline suggestions thing.

> Another corollary of this behaving like calling ``Resolve<B>()`` is that **Lifetime scopes are respected** using this relationship type. If you register an object as ``InstancePerDependency()`` and call the ``Func<B>`` multiple times, you'll get a new instance each time. However, if you register an object as ``SingleInstance()`` and call the ``Func<B>`` to resolve the object more than once, you will get *the same object instance every time*.

Author: Splitting existing suggestions across multiple lines
```suggestion
Further, lifetime scopes are respected. If you register an object as ``InstancePerDependency()`` and call the ``Func<B>`` multiple times, you'll get a new instance each time; if you register an object as ``SingleInstance()`` and call the ``Func<B>`` to resolve the object more than once, you will get *the same object instance every time*.
```

Reviewer: Yup yup.

An example of this relationship looks like:

.. sourcecode:: csharp
    public class B
    {
      public B() {}
      
      public void DoSomething() {}
    }

    public class A
    {
      Func<B> _newB;

      public A(Func<B> b) { _newB = b; }

      public void M()
      {
          var b = _newB();
          b.DoSomething();
      }
    }

Parameterized Instantiation (Func<X, Y, B>)
-------------------------------------------

> You can also use an *auto-generated factory* to provide parameters when creating an new instance of the object, where the constructor of the object calls for some additional parameters. As in the case of Dynamic Instantiation above, we're still doing something equivalent to calling ``.Resolve<B>()``, but in this case we're passing strongly-typed parameters to the resolution function. This is an alternative to :doc:`passing parameters during registration <../register/parameters>` or :doc:`passing during manual resolution <../resolve/parameters>`:

Reviewer: Maybe:

```suggestion
You can also use an *auto-generated factory* to provide parameters when creating an new instance of the object, where the constructor of the object calls for some additional parameters. While the ``Func<B>`` relationship is similar to ``Resolve<B>()``, the ``Func<X, Y, B>`` relationship is like calling ``Resolve<B>(TypedParameter.From<X>(x), TypedParameter.From<Y>(y))`` - a resolve operation that has typed parameters. This is an alternative to :doc:`passing parameters during registration <../register/parameters>` or :doc:`passing during manual resolution <../resolve/parameters>`:
```

.. sourcecode:: csharp
    public class B
    {
      public B(string someString, int id) {}
      
      public void DoSomething() {}
    }

    public class A
    {
        Func<int, string, B> _newB;

        public A(Func<int, string, B> b) { _newB = b }

        public void M()
        {
            var b = _newB(42, "http://hell.owor.ld");
            b.DoSomething();
        }
    }

> Note that since we're ``.Resolve()``ing the instantiation, rather than actually directly calling the constructor we don't need to declare the parameters in the same order as they appear in the ctor, nor do we necessarily need to provide *all* the parameters listed in that constructor. If some of the constructor's parameters could have been resolved by the Autofac Container already, then those parameters can be omitted from the ``Func`` signature being declared, so you only need to list the types that the Container can't resolve.

Reviewer: Inline code attached to other things doesn't really render well. Trying to unify references to the function type so it's consistent across all of them (`Func<B>` or `Func<X, Y, B>` rather than also having `Func`). Using "lifetime scope" instead of "container" to drive home that lifetime scope is obeyed. (Registrations can be added to the lifetime scope that don't exist in the container.)

```suggestion
Note that since we're resolving the instance rather than directly calling the constructor we don't need to declare the parameters in the same order as they appear in the constructor definition, nor do we need to provide *all* the parameters listed in the constructor. If some of the constructor's parameters can be resolved by the lifetime scope those parameters can be omitted from the ``Func<X, Y, B>`` signature being declared. You only need to list the types that the scope can't resolve.
```

Author: Note `If [...] parameters can be resolved by the lifetime scope those parameters can be omitted [...]` still benefits from the `, then` in the middle to make the sentence easier to parse.

```suggestion
Note that since we're resolving the instance rather than directly calling the constructor we don't need to declare the parameters in the same order as they appear in the constructor definition, nor do we need to provide *all* the parameters listed in the constructor. If some of the constructor's parameters can be resolved by the lifetime scope, then those parameters can be omitted from the ``Func<X, Y, B>`` signature being declared. You only *need* to list the types that the scope can't resolve.
```

Happy?

Reviewer: Yep. Looks good.

Alternatively, you can use this approach to override a constructor parameter that *would* otherwise have been resolved from the container, with a concrete instance already in hand.

Example:
.. sourcecode:: csharp
    //Suppose that P, Q & R are all registered with the Autofac Container.
    public class B
    {
      public B(int id, P peaDependency, Q queueDependency, R ourDependency) {}
      
      public void DoSomething() {}
    }

    public class A
    {
        Func<int, P, B> _newB;

        public A(Func<int, P, B> bFactory) { _newB = bFactory }

        public void M(P existingPea)
        {

>             var b = _newB(42, existingPea); //the needed Q and R will be resolved by Autofac, but not the P since the existingPea will be used.

Reviewer: Lines get a little long with inline comments.

```suggestion
            // The Q and R will be resolved by Autofac, but P will be existingPea instead.
            var b = _newB(42, existingPea);
```

            b.DoSomething();
        }
    }

> Internally, Autofac determines what values to use for the constructor args, solely based on the types, and behaves as though we've temporarily defined the input values as the objects that should be used to resolve those types. A direct consequence of this is that  **auto-generated function factories cannot have duplicate types in the input parameter list.** See below for further notes on this.

Reviewer: ```suggestion
Internally, Autofac determines what values to use for the constructor args solely based on the type and behaves as though we've temporarily defined the input values for resolution. A consequence of this is that  **auto-generated function factories cannot have duplicate types in the input parameter list.** See below for further notes on this.
```

Author: Just to check, supposing we overode a type that could otherwise be resolved, but that type was also being resolved elsewhere in the dependency chain, as a child dependency of some other auto-resolved ctor parameter, what would happen?

i.e. imagine:

```
    class A
    {
        public A(B b, C c)
        {
            CapturedBee = b;
            CapturedSea = c;
        }
        public B CapturedBee;
        public C CapturedSea;
    }

    class B
    {
        public B(C c)
        {
            CapturedSubSea= c;
        }
        public C CapturedSubSea;
    }

    class C
    {
        public C() {}
    }

    class Elsewhere
    {
        private Func<C, A> _newA;
        public A(Func<C, A> aFactory) { _newA = aFactory; }

        public void DoStuff(C existingSea)
        {
            var a = _newA(existingSea);

            var directResolution //true.
                = Object.ReferenceEquals(existingSea, a.CapturedSea);

            var indirectResolution // ???
                = Object.ReferenceEquals(existingSea, a.CapturedBee.CapturedSubSea);
        }
    }
```
What will `indirectResolution` be?

Reviewer: There's a lot of "it depends" because we don't know where the `existingSea` in `DoStuff` came from and we don't know the lifetime scopes in which A, B, and C are registered. If they're all singletons the answer may be different than if they're all instance-per-dependency.

However, parameters are provided to the immediate component being resolved, _not to things down the resolution chain_. [That's actually one of our FAQs.](https://autofac.readthedocs.io/en/latest/faq/injecting-global-parameters.html) For all intents and purposes, the `existingSea` and `capturedSea` are likely not going to be the same. They wouldn't be in any case, whether you were calling `Resolve` directly with parameters or using `Func` relationships.

Author: Cool. So it's not that you're temporarily overriding the *general* rules for resolving that type. It's specifically overriding *how it should be resolved for the argument of this **particular** ctor*

> 

Reviewer: Remove extra whitespace line.

```suggestion
```

> **Lifetime scopes are respected** using this relationship type, just as they are when using delegate factories. If you register an object as ``InstancePerDependency()`` and call the ``Func<X, Y, B>`` multiple times, you'll get a new instance each time. However, if you register an object as ``SingleInstance()`` and call the ``Func<X, Y, B>`` to resolve the object more than once, you will get *the same object instance every time regardless of the different parameters you pass in.* Just passing different parameters will not break the respect for the lifetime scope.

Reviewer: By moving this further away from the discussion about delegate factories, the "just as they are..." part gets disconnected. We need to link to add that context back.

```suggestion
**Lifetime scopes are respected** using this relationship type, just as they are when using :doc:`delegate factories <../advanced/delegate-factories>`. If you register an object as ``InstancePerDependency()`` and call the ``Func<X, Y, B>`` multiple times, you'll get a new instance each time. However, if you register an object as ``SingleInstance()`` and call the ``Func<X, Y, B>`` to resolve the object more than once, you will get *the same object instance every time regardless of the different parameters you pass in.* Just passing different parameters will not break the respect for the lifetime scope.
```

> Parameterized Instantiation on types with duplicate Types in their constructor args.

Author: Split existing Suggestion to cover all the lines.
```suggestion
```

> -------------------------------------------

Reviewer: Remove this header - it shouldn't be a whole new section that shows up in the table of contents. This is another where it appears a suggestion over multiple lines doesn't work.

```suggestion
```

Author: FBM. Created amd accepted single-line suggestions to reproduce this. So resolving this comment.

> -------------------------------------------

Author: Split existing Suggestion to cover all the lines.
```suggestion
```

> As noted above, a consequence of Parameterised Instantiation treating its args as TypedParameters is that **you cannot effectively use auto-generated function factories to invoke constructors which have duplicate types in their parameter list.** For example, say you have a type like this:

Reviewer: It feels redundant to have this here when it's basically a copy/paste of text we just read two paragraphs ago.

```suggestion
As noted above, ``Func<X, Y, B>`` treats arguments as ``TypedParameter`` so you can't have duplicate types in the parameter list. For example, say you have a type like this:
```

Author: Happy with the change around the "a consequence of ... is" clause.

But I was keen on the 2nd alternative phrasing of the bold section, as it's tying it back to the actual constructors.

It's making the point that "it's not really a problem with the Func<...> signature; it's a problem with the constructor signature. If you've got duplicate types in the constructor, then that's just not very compatible with this usage."

Thoughts?
I've made my case, so if it doesn't sway you then I'm happy to just apply this Suggestion as-is :)

Reviewer: I'd prefer not to over-bold stuff and, when we do, make the point once in bold and let folks continue reading. Two nearly-identical bold items are going to make the "skimmers" think nothing new was said.

Author: Fair. So remove the bolding of the 2nd version, but keep the words?

Reviewer: No, I think I prefer my suggestion. Thanks!

Author: no problem. Committed.
